transport/errors.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"fmt"
+	"strings"
+)
+
+// AsHandlerError converts an error into a BadRequestError or UnexpectedError,
+// leaving it unchanged if it's already one of the two.
+func AsHandlerError(err error) error {
+	if err == nil {
+		return err
+	}
+
+	switch err.(type) {
+	case BadRequestError, UnexpectedError:
+		return err
+	default:
+		return UnexpectedError{Reason: err}
+	}
+}
+
+// BadRequestError is a failure to process a request because the request was
+// invalid.
+type BadRequestError struct {
+	Reason error
+}
+
+func (e BadRequestError) Error() string {
+	return "BadRequest: " + e.Reason.Error()
+	// TODO were we planning on dropping these prefixes?
+}
+
+// UnexpectedError is a failure to process a request for an unexpected reason.
+type UnexpectedError struct {
+	Reason error
+}
+
+func (e UnexpectedError) Error() string {
+	return "Unexpected: " + e.Reason.Error()
+	// TODO were we planning on dropping these prefixes?
+}
+
+// MissingParametersError is a failure to process a request because it was
+// missing required parameters.
+type MissingParametersError struct {
```

Comment:
A follow up diff will flesh this out a bit more but the general idea is to do this,

```go
validator := RequestValidator{..}
// ..
request, err := validator.Validate()
if err != nil {
    return BadRequestError{err}
}

response, err := handle(request)
if err != nil {
    return UnexpectedError{err}
}
```

We don't want the prefix to always be added because if the error occurs while making an outgoing request, it'll be the plain error rather than being wrapped in `BadRequestError` or `UnexpectedError`.